README.md
# flutter_menu

This is a responsive UI framework developed for Web and Desktop. 

Try it out: http://www.flutter.psychtech.mitspace.dk/

The package gives you a desktop like experience on web (responsive to touch and desktop displays). It gives you the following:

![Example](screenshots/example6.gif)

- [x] Menu with list items (with icon, title and keyboard shortcuts)
- [x] Keyboard shortcuts for menu items
- [x] Master, detail views (Left pane, right pane)
- [x] Pane sizes can be changed by user (or not - you choose).
- [x] Context menu (right click) for Master, Detail and widgets
- [x] Longpress for contextmenu (for touch displays). Centered and animated
- [x] Choose between touch UI or desktop UI (can be change by user)
- [x] Drawer (small or large)
- [x] Pane sizes is set by a flex factor for the UI to stay nice and responsive.


- [ ] different menu in compact mode (to be responsive)
- [ ] More predefined contextmenus to choose from
- [ ] Scrollbar
- [ ] More callbacks on change
- [ ] Extra topbar to be used as toolbar or information bar
- [ ] Fullscreen dialogs

All siblings down the widget tree has access to the AppScreen functionality.
 
This is a beta release and breaking changes will come soon!

Get support, announcements and share knowledge: https://discord.gg/pmXCFZUxxC

## How to install

### On web:

The Flutter App has to have control over the right click (contextmenu).
To take control you have to include the following in your index.html file in the web folder:

```html
<body oncontextmenu="return false;"></body>
```

### On desktop:

Nothing has to be changed.


## How to use

You can access all variables and functions through the buildContext, and for your convenience we have made an extention for easy access:

```dart
context.appScreen
```

the AppScreen has to be in a parent BuildContext (and not the current context). This is why panes has to be a Builder function and not just a Widget. To prevent you from making misconceptional errors.

## Menus

The menu can have several menuItems and each menuItem has its own menuListItems.

Each menuItem decides width and height of the menuItemList widget and can be set as active or not.

MenuListItem can be a MenuListDivider() or a MenuListItem.

MenuListItems can have icons, shortcuts and have a callback function: onPressed().

See example.

Programmably you can control:

```dart
context.appScreen.hideMenu(); // hides the menu from the screen. Keyboard shortcuts is still active
context.appScreen.showMenu(); // shows the menu on the screen
context.appScreen.openMenu(); // opens the active menuitem
context.appScreen.closeMenu(); // close a open menuitem (if any)
```


You have access to current variables:

```dart
context.appScreen.isMenuShown(); // returns true if menu is shown on Screen
context.appScreen.isMenuOpen();  // returns true if a MenuItem is shown on Screen
```


## Keyboard shortcuts

It is possible to activate functions with keyboard shortcuts which is good for productibility.

Each MenListItem() can be supplied with a MenuShortcut().

*MenuShortcut.key* is of type LogicalKeyboardKey and is defined in flutter/services.dart. This should be a normal physical key. 

You can then choose which system keys has to be hold down as well: ctrl, alt, shift.
Please notice that shift should only be used in combination with other system keys and notice that the browser takes presidence over the app for keyboard shortcuts - so choose with care.


## Keyboard Shortcut Overlay

Overlay feature is disabled by default, but can be enabled programmably (se How-to-use section). This can be helpfull under development and as a feature making example videos.

```dart
context.appScreen.showShortcutOverlay();  // 2 sec. text overlay will be shown
context.appScreen.hideShortcutOverlay();  // No overlay will be shown
```


## Master/ detail panes

You always have to set the Master pane. If no detail pane is supplied only Master is shown.
If you set both Master and detail panes default behavior will be give the user control of a moveable vertical line in the UI to change the size of the detail pane.

In AppScreen() you can set:

```dart
  final double masterPaneFlex;      // default=1
  final double detailPaneFlex;      // default=1
  final double masterPaneMinWidth;  // default= 500
  final double detailPaneMinWidth;  // default = 500
```

The flex sizing is just as in the Expanded() widget. You set the relative size, as for the UI to stay responsive when window size changes. The UI updates with respect for the minimum sizes, and if the detailPane can not stay on screen, only master will be shown.

The user can change the sizes of the master and detail panes, if this is possible within the minium size boundaries.


The Master/detail panes are responsive, and you can set the breakpoint:

*desktopBreakpoint (=800)* if screen width is under breakpoint value the compact view is shown
*onBreakPointChange* A callback function to update your UI on change (it is automatically delayed to next tick to prevent call under UI build).

You have access to current variables:

*context.appScreen.isDesktop()* - returns true if Desktop and false if compact
*context.appScreen.detailIsShown()* - returns true if detail is shown

You can get the screen and pane sizes and topleft and bottomright corner position:

```dart
context.appScreen.screenDetails
context.masterPaneDetails
context.master
```
```dart
class Detail {
  double height;
  double width;
  double minDx;
  double minDy;
  double maxDx;
  double maxDy;
}
```

In compact mode you can change master/detail view:

```dart
context.appScreen.showOnlyMaster(); //  change UI to show masterPane()
context.appScreen.showOnlyDetail(); // change UI to show detailPane() 
```

## Drawer

The AppScreen has its own drawer, just as the Scaffold() has.

You can choose set only one size of drawer or you can have to different sizes for the user to choose.

```dart
class AppDrawer {
  final double smallDrawerWidth;
  final Widget smallDrawer;
  final double largeDrawerWidth;
  final Widget largeDrawer;
  final bool showOnDesktop;
  final bool defaultSmall;
  final bool autoSizing;            // autoSizing is not yet implemented.
  final double edgeDragOpenWidth;
}
```
*default small* if true Drawer opens with the small drawer on start.

*showOnDesktop* if true Drawer will be shown automaticly on Desktop is AppScreen is big enough.

*edgeDragOpenWith* is for touch users to slide the drawer open. It indicates the zone where dragging is registred. On mobile devices it is set to 20 in the Drawer() of Scaffold(). But in browsers run on mobile devices there can be a sliding function the change webpage. The zone has been made bigger to adjust for this. You might change it according to your needs. If set to 0 drag to Open Drawer will be disabled.

## Context menu (right click or longpress)

You can set different context menus for Master & Detail pane, and each individual Widget can have its own context menu.

If contextmenu is activated by a longpress it will be centered and animated. If it is right clicked it will be top left position and no animation.

```dart
AppScreen(
  masterContextMenu: ContextMenu(
    width: 150,
    height: 250,
    child: ContextMenuSliver(
      title: 'Master',
      children: [
        masterContextMenuItem(color: 'Red'),
        masterContextMenuItem(color: 'Blue'),
        masterContextMenuItem(color: 'Purple'),
        masterContextMenuItem(color: 'Pink'),
      ],
    ),
  ),
)
```
This gives us a ContextMenu as shown in the masterPane (it stays within boundaries of the Master Pane).


You just wrap a widget in a ContextMenu() widget and specify the widget. You can use the prebuild ContextMenuSliver() or build your own. 

The only requirement is that the ContextMenu() is a sibling to AppScreen(). 

```dart
ContextMenuContainer(
    menu: ContextMenuSliver(title: 'Widget'),
    menuWidth: 200,
    menuHeight: 200,
    child: SizedBox(height: 200, width: 200),
)
```

This will give you a small empty ContextMenu. You can use ContextMenuSliverItem to build your menu.

Please take a look at the example app for more examples.

More prebuild ContextMenus will come. Feel free to build your own and contribute to the project.



CHANGELOG.md
# [0.4.1] - 2021-05-14

* BUG FIX: Issue 9: Renderflex overflow on single pane to double pane change 

# [0.4.0] - 2021-05-13

* Updated to nullsafety

# [0.3.2] - 2021-02-28

* Example updated

# [0.3.1] - 2020-11-13

* New drawer example
* new gif video for readme.md

# [0.3.0] - 2020-11-12

* FEATURE:  Drawer added (top left menubar icon to change or touch dragging enabled)
 

# [0.2.3] - 2020-11-09

* CHANGE: ResizeBarHelper only in touchMode
* CHANGE: Options to set touchMenuBar and desktopMenuBar height in AppScreen()
* CHANGE: Bigger textsize in touchmode


# [0.2.2] - 2020-11-09

* FEATURE: New ResizeBar() to customize the look of the resizeBar. Touch screens can now use the circular helper to move it around
           use leftColor and RightColor to get at default BoxDecoration or customize with decoration and helperDecoration.
* FEATURE: First version of touch mode. Get a larger menu for easy touch screen use.

# [0.2.1] - 2020-11-09

* BREAKING: Pane sizing has changed: You specifiy flex for Master and Detail and minimum sizes for Master and Detail. desktopBreakpoint has to be equal to or larger than min sizes for Master + Detail.
* FEATURE: Longpress opens a centered and animated contextmenu (right click topleft positioned and no animation)

# [0.1.2] - 2020-11-07

* Animation added to context menu
* Context menu stay inside boundaries of Master or Detail pane
* BREAKING: masterContextMenu and detailContextMenu of new type ContextMenu
* BREAKING: ContextMenu (for custom widget menus) is now ContextMenuContainer.
* BREAKING: ContextMenus has to specify width and height (new feature: contextmenu stays within pane and new center option )

# [0.1.1] - 2020-11-04

* Bug fix: when contextmenu then open left click would remove focus.

# [0.1.0] - 2020-11-03

* First version with contextmenu (master, detail and custom widget)


# [0.0.4] - 2020-10-31

* Example gif updated

# [0.0.3] - 2020-10-31

* News: responsive master/detail pane included with lots of features
* Breaking: Menu() changed to AppScreen()
* new example website: http://www.flutter.psychtech.mitspace.dk/
* new gif in readme

# [0.0.2] - 2020-10-25

* changes to readme - gif example included

# [0.0.1] - 2020-10-24

* initial release

lib/src/menu_model.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

/// MenuList
class MenuItem {
  final double width;
  final double height;

  final String title;
  final bool isActive;
  final List<MenuListItemType> menuListItems;

  const MenuItem({
    this.width = 170,
    this.height = 200,
    this.title = "Empty",
    this.isActive = true,
    required this.menuListItems,
  });

  @override
  String toString() =>
      'MenuItem(title: $title, isActive: $isActive, menuItemList: $menuListItems)';
}

/// Abstract class
abstract class MenuListItemType {}

/// Use to have a dividerline in your menulist
class MenuListDivider extends MenuListItemType {}

/// Menu Items for your menuList
class MenuListItem extends MenuListItemType {
  final String title;
  final IconData? icon;
  final bool isActive;
  final MenuShortcut? shortcut;
  final Function? onPressed;
  MenuListItem({
    required this.title,
    this.icon,
    this.isActive = true,
    this.shortcut,
    this.onPressed,
  });

  @override
  String toString() {
    return 'MenuListItem(title: $title, icon: $icon, isActive: $isActive, shortcut: $shortcut, onPressed: $onPressed)';
  }
}

/// Setup a keyboard shortcut for MenuItems, notice that shift should only be used in combination with other system keys and take care
/// that the browser takes presidence over the app for keyboard shortcuts so choose with care
class MenuShortcut {
  final LogicalKeyboardKey? key;
  final bool shift;
  final bool alt;
  final bool ctrl;

  MenuShortcut({
    this.key,
    this.shift = false,
    this.alt = false,
    this.ctrl = false,
  });

  @override
  String toString() {
    return 'MenuShortcut(logicalKey: $key, shift: $shift, alt: $alt, ctrl: $ctrl)';
  }
}

// /// DetailPane info:
// class DetailPane {
//   final double defaultWidth;
//   final double minWidth;
//   final double maxWidth;
//   final bool fixedWidth;

//   Pane(this.defaultWidth, this.minWidth, this.maxWidth);
// }


lib/src/contextmenu.dart
import 'package:flutter/material.dart';
import 'appscreen.dart';

enum ContextPlacementCenter { longPressOnly, never, always }

abstract class ContextMenuWidget extends StatelessWidget {}

abstract class ContextMenuItemType extends StatelessWidget {}

class ContextMenuContainer extends StatelessWidget {
  final ContextMenuWidget menu;
  final double width;
  final double height;
  final Widget? child;
  const ContextMenuContainer(
      {Key? key,
      required this.menu,
      required this.height,
      required this.width,
      this.child})
      : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Center(
      child: GestureDetector(
        onLongPressStart: (details) {
          context.appScreen.showContextMenu(
              offset: details.globalPosition,
              menu: menu,
              width: width,
              height: height,
              center: true);
        },
        child: Listener(
          behavior: HitTestBehavior.opaque,
          onPointerDown: (event) {
            if (event.buttons == 2) // right click
            {
              context.appScreen.showContextMenu(
                  offset: event.position,
                  menu: menu,
                  width: width,
                  height: height,
                  center: false);
            } else
              context.appScreen.hideContextMenu();
          },
          child: child,
        ),
      ),
    );
  }
}


lib/src/appscreen_model.dart
import 'package:flutter/material.dart';

import '../flutter_menu.dart';

class Detail {
  late double height;
  late double width;
  late double minDx;
  late double minDy;
  late double maxDx;
  late double maxDy;
  // bool isShown;
  @override
  String toString() =>
      'Detail(height: $height, width: $width, minDx: $minDx), minDy: $minDy, maxDx: $maxDx, minDy: $maxDy)';
}

/// ContextMenu
class ContextMenu {
  final double width;
  final double height;
  final ContextMenuWidget child;

  ContextMenu({required this.child, required this.width, required this.height});
}

/// Custom resizeBar: either use leftColor & rightcolor or decorations. Not both of them
class ResizeBar {
  final double width;
  Color? leftColor;
  Color? rightColor;
  Decoration? decoration;
  final double helperSize;
  Decoration? helperDecoration;

  final double helperPos;

  ResizeBar({
    this.width = 5,
    this.decoration,
    this.helperSize = 20,
    this.helperDecoration,
    this.leftColor,
    this.rightColor,
    this.helperPos = 30,
  });
}

ResizeBar kDefaultResizeBar = ResizeBar(
  width: 5,
  helperSize: 20,
  helperPos: 30,
  leftColor: Colors.white,
  rightColor: Colors.blueGrey,
);

class AppDrawer {
  final double smallDrawerWidth;
  final Widget? smallDrawer;
  final double largeDrawerWidth;
  final Widget? largeDrawer;
  final bool showOnDesktop;
  final bool defaultSmall;
  final bool autoSizing;
  final double edgeDragOpenWidth;
  const AppDrawer({
    this.smallDrawerWidth = 70,
    this.smallDrawer,
    this.largeDrawerWidth = 100,
    this.largeDrawer,
    this.showOnDesktop = true,

    /// Start with small (true) or large drawer (false)
    this.defaultSmall = true,

    /// When drawer is closed, the user can open it by sliding in the left side area
    this.edgeDragOpenWidth = 40,

    /// autosizing not yet implemented
    this.autoSizing = true,
  })  : assert(smallDrawerWidth >= 30,
            "Has to be at least 30 for UI to render nicely!"),
        assert(largeDrawerWidth >= 30,
            "Has to be at least 30 for UI to render nicely!"),
        assert(edgeDragOpenWidth >= 0, "Has to be at a positive number!");
}


lib/src/contextmenu_widgets.dart
import 'package:flutter/material.dart';

import 'package:flutter_menu/flutter_menu.dart';

class ContextMenuSliver extends ContextMenuWidget {
  final String title;
  final double height;
  final double width;
  final Color titleBackgroundColor;
  final TextStyle? titleStyle;
  final Color widgetBackgroundColor;

  final List<Widget> children;

  ContextMenuSliver({
    Key? key,
    this.title = 'Menu',
    this.height = 200,
    this.width = 150,
    this.titleBackgroundColor = Colors.green,
    this.titleStyle,
    this.widgetBackgroundColor = Colors.amber,
    this.children = const [Center(child: Text('Empty'))],
  });

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      height: height,
      width: width,
      child: Card(
        elevation: 5,
        child: Container(
          color: widgetBackgroundColor,
          child: CustomScrollView(
            slivers: <Widget>[
              SliverAppBar(
                title: Text(title, style: titleStyle),
                backgroundColor: titleBackgroundColor,
                expandedHeight: 60.0,
                // flexibleSpace: FlexibleSpaceBar(
                //   background: Image.asset('assets/forest.jpg', fit: BoxFit.cover),
                // ),
              ),
              SliverFixedExtentList(
                itemExtent: 50.0,
                delegate: SliverChildListDelegate(List.from(children)),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class ContextMenuItem extends ContextMenuItemType {
  final Function? onTap;

  final Widget? child;

  ContextMenuItem({
    Key? key,
    this.onTap,
    this.child,
  });
  @override
  Widget build(BuildContext context) {
    return InkWell(onTap: onTap as void Function()?, child: child);
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_menu/flutter_menu.dart';

const Map kColorMap = {
  'Red': Colors.red,
  'Blue': Colors.blue,
  'Purple': Colors.purple,
  'Black': Colors.black,
  'Pink': Colors.pink,
  'Yellow': Colors.yellow,
  'Orange': Colors.orange,
  'White': Colors.white,
  'BlueGrey': Colors.blueGrey,
};

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        textButtonTheme: TextButtonThemeData(
          style: ButtonStyle(
            overlayColor: MaterialStateProperty.all(
                Colors.blueGrey[600]), // Set Button hover color
          ),
        ),
        elevatedButtonTheme: ElevatedButtonThemeData(
          style: ButtonStyle(
            foregroundColor: MaterialStateProperty.all(Colors.black),

            backgroundColor: MaterialStateProperty.all(Colors.white),
            overlayColor: MaterialStateProperty.all(
                Colors.blueGrey[600]), // Set Button hover color
          ),
        ),
      ),
      debugShowCheckedModeBanner: false,
      home: Screen(),
    );
  }
}

class Screen extends StatefulWidget {
  @override
  _ScreenState createState() => _ScreenState();
}

class _ScreenState extends State<Screen> {
  final ScrollController scrollController = ScrollController();
  TextEditingController controller = TextEditingController();
  String _message = "Choose a MenuItem.";
  String _drawerTitle = 'Tap a drawerItem';
  IconData _drawerIcon = Icons.menu;

  Color masterBackgroundColor = Colors.white;
  Color detailBackgroundColor = Colors.blueGrey[300] as Color;

  void _showMessage(String newMessage) {
    setState(() {
      _message = newMessage;
    });
  }

  void _masterSetBackgroundColor(String color) {
    setState(() {
      masterBackgroundColor = kColorMap[color];
    });
  }

  void _detailSetBackgroundColor(String color) {
    setState(() {
      detailBackgroundColor = kColorMap[color];
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: AppScreen(
        masterContextMenu: ContextMenu(
          width: 150,
          height: 250,
          child: ContextMenuSliver(
            title: 'Master',
            children: [
              masterContextMenuItem(color: 'Red'),
              masterContextMenuItem(color: 'Blue'),
              masterContextMenuItem(color: 'Purple'),
              masterContextMenuItem(color: 'Pink'),
              masterContextMenuItem(color: 'White'),
            ],
          ),
        ),
        detailContextMenu: ContextMenu(
          width: 300,
          height: 150,
          child: ContextMenuSliver(
            title: 'Detail',
            children: [
              detailContextMenuItem(color: 'Yellow'),
              detailContextMenuItem(color: 'Orange'),
              detailContextMenuItem(color: 'Pink'),
              detailContextMenuItem(color: 'Red'),
              detailContextMenuItem(color: 'BlueGrey'),
            ],
          ),
        ),
        menuList: [
          MenuItem(title: 'File', menuListItems: [
            MenuListItem(
              icon: Icons.open_in_new,
              title: 'Open',
              onPressed: () {
                _showMessage('File.open');
              },
              shortcut: MenuShortcut(key: LogicalKeyboardKey.keyO, ctrl: true),
            ),
            MenuListItem(
              title: 'Close',
              onPressed: () {
                _showMessage('File.close');
              },
            ),
            MenuListItem(
              title: 'Save',
              onPressed: () {
                _showMessage('File.save');
              },
            ),
            MenuListItem(
              title: 'Delete',
              shortcut: MenuShortcut(key: LogicalKeyboardKey.keyD, alt: true),
              onPressed: () {
                _showMessage('File.delete');
              },
            ),
          ]),
          MenuItem(title: 'View', isActive: true, menuListItems: [
            MenuListItem(title: 'View all'),
            MenuListItem(title: 'close view'),
            MenuListItem(title: 'jump to'),
            MenuListItem(title: 'go to'),
          ]),
          MenuItem(title: 'Help', isActive: true, menuListItems: [
            MenuListItem(title: 'Help'),
            MenuListItem(title: 'About'),
            MenuListItem(title: 'License'),
            MenuListDivider(),
            MenuListItem(title: 'Goodbye'),
          ]),
        ],
        masterPane: masterPane(),
        detailPane: detailPane(),
        drawer: AppDrawer(
          defaultSmall: false,
          largeDrawerWidth: 200,
          largeDrawer: drawer(small: false),
          smallDrawerWidth: 60,
          smallDrawer: drawer(small: true),
        ),
        onBreakpointChange: () {
          setState(() {
            print('Breakpoint change');
          });
        },
        resizeBar: ResizeBar(
            leftColor: masterBackgroundColor,
            rightColor: detailBackgroundColor),
      ),
    );
  }

  Widget drawer({required bool small}) {
    return Container(
        color: Colors.amber,
        child: ListView(
          children: [
            drawerButton(
                title: 'User', icon: Icons.account_circle, small: small),
            drawerButton(title: 'Inbox', icon: Icons.inbox, small: small),
            drawerButton(title: 'Files', icon: Icons.save, small: small),
            drawerButton(
              title: 'Clients',
              icon: Icons.supervised_user_circle,
              small: small,
            ),
            drawerButton(
              title: 'Settings',
              icon: Icons.settings,
              small: small,
            ),
          ],
        ));
  }

  Widget drawerButton(
      {required String title, required IconData icon, required bool small}) {
    return small
        ? drawerSmallButton(icon: icon, title: title)
        : drawerLargeButton(icon: icon, title: title);
  }

  Widget drawerLargeButton({required String title, required IconData icon}) {
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: Card(
          elevation: 8,
          child: ListTile(
            leading: Icon(icon),
            title: Text(title),
            onTap: () {
              setState(() {
                _drawerIcon = icon;
                _drawerTitle = title;
              });
            },
          )),
    );
  }

  Widget drawerSmallButton({required String title, required IconData icon}) {
    return Padding(
      padding: const EdgeInsets.fromLTRB(3, 8, 3, 8),
      child: Card(
          elevation: 8,
          child: SizedBox(
            height: 40,
            child: ElevatedButton(
              onPressed: () {
                setState(() {
                  _drawerIcon = icon;
                  _drawerTitle = title;
                });
              },
              child: Center(child: Icon(icon, size: 30, color: Colors.black54)),
            ),
          )),
    );
  }

  Builder detailPane() {
    print('BUILD: detailPane');
    return Builder(
      builder: (BuildContext context) {
        return Container(
          color: detailBackgroundColor,
          child: SingleChildScrollView(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                SizedBox(height: 20),
                Card(
                  elevation: 12,
                  child: Container(
                    width: 300,
                    height: 50,
                    child: Container(
                      color: Colors.amber,
                      child: Center(
                          child:
                              Text('DETAIL', style: TextStyle(fontSize: 20))),
                    ),
                  ),
                ),
                SizedBox(height: 20),
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceAround,
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: [
                    ElevatedButton(
                      onPressed: () {
                        context.appScreen.closeMenu();
                      },
                      child: Text('Close Menu'),
                    ),
                  ],
                ),
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceAround,
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: [
                    ElevatedButton(
                      onPressed: () {
                        context.appScreen.hideMenu();
                      },
                      child: Text('Hide Menu'),
                    ),
                    ElevatedButton(
                      onPressed: () {
                        context.appScreen.showMenu();
                      },
                      child: Text('Show Menu'),
                    ),
                  ],
                ),
                SizedBox(height: 20),
                SizedBox(
                  width: 300,
                  height: 300,
                  child: Container(
                    color: Colors.blueGrey,
                    child: Center(
                        child: Text(_message,
                            textAlign: TextAlign.center,
                            style: TextStyle(fontSize: 40))),
                  ),
                ),
                SizedBox(height: 20),
                Card(
                  elevation: 12,
                  child: Container(
                    width: 300,
                    height: 50,
                    child: Center(
                      child: Text(
                          'Pane height: ${context.appScreen.detailPaneDetails.height.toStringAsFixed(1)} width: ${context.appScreen.detailPaneDetails.width.toStringAsFixed(1)}',
                          style: TextStyle(fontSize: 20)),
                    ),
                  ),
                ),
                SizedBox(height: 20),
                if (context.appScreen.isCompact())
                  ElevatedButton(
                    onPressed: () {
                      context.appScreen.showOnlyMaster();
                    },
                    child: Text('Show master'),
                  ),
              ],
            ),
          ),
        );
      },
    );
  }

  Builder masterPane() {
    print('BUILD: masterPane');
    return Builder(
      builder: (BuildContext context) {
        return Container(
          color: masterBackgroundColor,
          child: SingleChildScrollView(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                SizedBox(height: 20),
                Card(
                  elevation: 12,
                  child: Container(
                    width: 300,
                    height: 50,
                    child: Container(
                      color: Colors.amber,
                      child: Center(
                          child:
                              Text('MASTER', style: TextStyle(fontSize: 20))),
                    ),
                  ),
                ),
                SizedBox(height: 80),
                SizedBox(
                    width: 140,
                    height: 110,
                    child: Card(
                        elevation: 8,
                        child: Column(
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: [
                            Icon(_drawerIcon, size: 50, color: Colors.black54),
                            Text(_drawerTitle,
                                style: TextStyle(color: Colors.black54)),
                          ],
                        ))),
                SizedBox(height: 80),
                ContextMenuContainer(
                  width: 300,
                  height: 200,
                  menu: ContextMenuSliver(
                    title: 'Widget',
                  ),
                  child: SizedBox(
                    width: 300,
                    height: 300,
                    child: Container(
                      color: Colors.blueGrey,
                      child: Center(
                        child: Text('Right click or longpress me',
                            textAlign: TextAlign.center,
                            softWrap: true,
                            style: TextStyle(fontSize: 30)),
                      ),
                    ),
                  ),
                ),
                SizedBox(height: 20),
                if (context.appScreen.isCompact())
                  ElevatedButton(
                    onPressed: () {
                      context.appScreen.showOnlyDetail();
                    },
                    child: Text('Show detail'),
                  ),
              ],
            ),
          ),
        );
      },
    );
  }

  Builder appContextMenu() {
    print('BUILD: appContextMenu');
    return Builder(
      builder: (BuildContext context) {
        return SizedBox(
          height: 300,
          width: 400,
          child: Container(
            color: Colors.yellow,
            child: Text('AppContextMenu'),
          ),
        );
      },
    );
  }

  Widget masterContextMenuItem({required String color}) {
    return ContextMenuItem(
      onTap: () {
        _masterSetBackgroundColor(color);
      },
      child: Container(
        color: kColorMap[color],
        child: Center(child: Text(color)),
      ),
    );
  }

  Widget detailContextMenuItem({required String color}) {
    return ContextMenuItem(
      onTap: () {
        _detailSetBackgroundColor(color);
      },
      child: Container(
        color: kColorMap[color],
        child: Center(child: Text(color)),
      ),
    );
  }
}
